### PR TITLE
Use thread-sanitizer-compatible fencing

### PR DESCRIPTION
**Pull request by thomcc**
This also disables the x86 asm under tsan.

### CODE REVIEW

File: src/rwlock.rs [hunk 1]
```diff
@@ -1,12 +1,15 @@
 use super::shared::{SpinWait, Waiter};
-#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[cfg(all(
+    not(usync_tsan_enabled),
+    any(target_arch = "x86", target_arch = "x86_64"),
+))]
 use lock_api::RawRwLock as _RawRwLock;
```
Comment (kprotty) [inline src/rwlock.rs]: Could this use statement be moved into `try_lock_exclusive_assuming` for the x86 specialization?

File: src/rwlock.rs [hunk 2]
```diff
@@ -125,7 +128,10 @@ unsafe impl lock_api::RawRwLock for RawRwLock {
 
 //  --- X86 Specializations
 
-#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[cfg(all(
+    not(usync_tsan_enabled),
```
Comment (kprotty) [inline src/rwlock.rs]: Could this be removed, then change the x86 asm to use the tsan-friendly fence instead?

For example
```rs
let acquired = old_locked_bit == 0;
if acquired {
    fence_acquire(&self.state);
}

acquired
```
This clears up the TSAN errors on x86 for my side.

**Review by kprotty — CHANGES_REQUESTED** [review]

File: src/rwlock.rs (hunk 2 above)
Comment (thomcc) [reply src/rwlock.rs]: Oh, yeah, that should work.

**Comment by thomcc** [comment]
This passes on x86 under tsan for me as well, aside from doctests (which seems like a rustc bug).

**Comment by thomcc** [comment]
The CI error is unrelated. One of the changes in #5 fixes it (but it also turns on a bunch of tests that are mad for reasons I haven't looked into yet).

**Review by kprotty — APPROVED** [review]
Nice

**Pull request merged**